app/bundles/CategoryBundle/Model/ContactActionModel.php

```diff
@@ -0,0 +1,68 @@
+<?php
+
+/*
+ * @copyright   2018 Mautic Contributors. All rights reserved
+ * @author      Mautic
+ *
+ * @link        http://mautic.org
+ *
+ * @license     GNU/GPLv3 http://www.gnu.org/licenses/gpl-3.0.html
+ */
+
+namespace Mautic\CategoryBundle\Model;
+
+use Mautic\LeadBundle\Model\LeadModel;
+
+class ContactActionModel
+{
+    /**
+     * @var LeadModel
+     */
+    private $contactModel;
+
+    /**
+     * @param LeadModel $contactModel
+     */
+    public function __construct(LeadModel $contactModel)
+    {
+        $this->contactModel = $contactModel;
+    }
+
+    /**
+     * @param array $contactIds
+     * @param array $categoryIds
+     */
+    public function addContactsToCategories(array $contactIds, array $categoryIds)
+    {
+        $contacts = $this->contactModel->getLeadsByIds($contactIds);
+
+        foreach ($contacts as $contact) {
+            if (!$this->contactModel->canEditContact($contact)) {
+                continue;
+            }
+
+            $this->contactModel->addToCategory($contact, $categoryIds);
+            $this->contactModel->detachEntity($contact);
```

Comment:
Clearing memory. Any problem that it's there?

Reply:
Do we have any reason for this? We will call both methods `addContactsToCategories` and `removeContactsFromCategories`. I think this would cause EM need to hydrate same entity more times, but really not sure. Actually I am not sure why to do this and advantage this approach has, I had never did / test this before.

Reply:
Good point, will remove and let Doctrine to deal with it

Reply:
Removed